### PR TITLE
Improve detection japanese region

### PR DESCRIPTION
**Pull request by alexiglesias31**
Rel card: https://app.clickup.com/t/85zu3b9q8

Reported issue:
- City may be empty in the order (this depends if the customer inputs the city in the Amazon address, it's not required for JP)
- Last name filled with "." (this also depends if the user inputs the fullname without spaces, like `テストアマゾン`, "." is the fallback, since last name is required for WC compatibility purposes)

Proposed fix:
- Use "-" instead of "." as a filler
- Improve detection of the Prefecture, using the same snippet than Stripe

### CODE REVIEW

**Review by Greg36 — APPROVED** [review]
Looks good!

File: includes/class-wc-amazon-payments-advanced-utils.php [hunk 1]
```diff
@@ -26,4 +26,43 @@ public static function get_edit_order_screen_id() {
...
+	 * @return array
+	 */
+	public static function get_non_required_fields_per_country() {
+
+		$non_required_fields = array(
+			'JP' => array(
+				'billing_city',
+				'shipping_city',
+				'billing-city',
+				'shipping-city',
```
Comment (lauramelos) [inline includes/class-wc-amazon-payments-advanced-utils.php:62]: why are reapeted here?

File: includes/class-wc-gateway-amazon-payments-advanced.php [hunk 2]
```diff
@@ -1036,10 +1037,12 @@ public function display_payment_method_selected( $checkout_session = null ) {
...
 	 *
 	 * @param  object $checkout_session The active checkout session.
+	 * @param  bool   $force            Whether to force the refresh of the checkout session.
+	 *
 	 * @return boolean
 	 */
-	public function has_payment_preferences( $checkout_session = null ) {
-		if ( null === $checkout_session ) {
+	public function has_payment_preferences( $checkout_session = null, $force = false ) {
+		if ( ! $force && null === $checkout_session ) {
```
Comment (lauramelos) [inline includes/class-wc-gateway-amazon-payments-advanced.php:1045]: if the $checkout_session is null shouldn't be refreshed? I mean the conditional should be an OR

File: includes/class-wc-gateway-amazon-payments-advanced.php [hunk 3]
```diff
@@ -1054,10 +1057,11 @@ public function has_payment_preferences( $checkout_session = null ) {
 	 * Returns the selected payment method from Amazon.
 	 *
 	 * @param object $checkout_session The active checkout session.
+	 * @param bool   $force            Whether to force the refresh of the checkout session.
 	 * @return string
 	 */
-	public function get_selected_payment_label( $checkout_session = null ) {
-		if ( null === $checkout_session ) {
+	public function get_selected_payment_label( $checkout_session = null, $force = false ) {
+		if ( ! $force && null === $checkout_session ) {
```
Comment (lauramelos) [inline includes/class-wc-gateway-amazon-payments-advanced.php:1064]: same here.

**Review by lauramelos — COMMENTED** [review]
Leave some comments

File: includes/class-wc-amazon-payments-advanced-utils.php (hunk 1 above)
Comment (alexiglesias31) [reply includes/class-wc-amazon-payments-advanced-utils.php]: Refactored this, we need the ones with '-' as separator for checkout blocks

File: includes/class-wc-gateway-amazon-payments-advanced.php (hunk 2 above)
Comment (alexiglesias31) [reply includes/class-wc-gateway-amazon-payments-advanced.php]: I was getting a kind of infinite loop in my local when a non-logged user opened the checkout page (Checkout Blocks).
I adjusted the parameter name, the idea is to have bypass creating a new session if the existing session is already null, in checkout blocks this was running over and over causing RAM usage issues.

File: includes/class-wc-amazon-payments-advanced-utils.php [hunk 4]
```diff
@@ -54,15 +54,23 @@ public static function get_non_required_fields() {
...
-		return apply_filters( 'woocommerce_amazon_pa_non_required_fields_per_country', $non_required_fields );
+		$non_required_fields = array();
+
+		foreach ( $mapped_fields_per_country as $country => $fields ) {
+			foreach ( $fields as $field ) {
+				$non_required_fields[ $country ][] = 'billing_' . $field;
+				$non_required_fields[ $country ][] = 'billing-' . $field;
+				$non_required_fields[ $country ][] = 'shipping_' . $field;
+				$non_required_fields[ $country ][] = 'shipping-' . $field;
+			}
```
Comment (lauramelos) [inline includes/class-wc-amazon-payments-advanced-utils.php:71]: Again here, why we have two billing and two shipping?

Comment (alexiglesias31) [reply includes/class-wc-amazon-payments-advanced-utils.php:71]: It's billing_city for legacy checkout and billing-city for blocks checkout

File: includes/class-wc-gateway-amazon-payments-advanced.php (hunk 2 above)
Comment (lauramelos) [reply includes/class-wc-gateway-amazon-payments-advanced.php]: I was missundertanding the idea.

**Review by lauramelos — APPROVED** [review]

File: includes/class-wc-gateway-amazon-payments-advanced.php (hunk 2 above)
Comment (alexiglesias31) [reply includes/class-wc-gateway-amazon-payments-advanced.php]: The description was far from clear 😞

**Comment by alexiglesias31** [comment]
@lauramelos adding some new commits to fix the following issues:
- https://app.clickup.com/t/86dtchqkr
- https://app.clickup.com/t/86dtwrkaw
- https://app.clickup.com/t/86dth8a9m

**Review by lauramelos — APPROVED** [review]
LGTM

**Review by lauramelos — APPROVED** [review]

**Pull request merged**